pennylane/plugins/default_qubit.py

```diff
@@ -345,35 +345,44 @@ def mat_vec_product(self, mat, vec, wires):
         return np.reshape(state_multi_index, 2 ** self.num_wires)
 
     def expval(self, observable, wires, par):
-        r"""Expectation value of observable on specified wires.
-
-        Args:
-          observable      (str): name of the observable
-          wires (Sequence[int]): target subsystems
-          par    (tuple[float]): parameter values
-
-        Returns:
-          float: expectation value :math:`\expect{A} = \bra{\psi}A\ket{\psi}`
-            """
-
-        A = self._get_operator_matrix(observable, par)
         if self.shots == 0:
             # exact expectation value
+            A = self._get_operator_matrix(observable, par)
             ev = self.ev(A, wires)
         else:
             # estimate the ev
-            # sample Bernoulli distribution n_eval times / binomial distribution once
-            a, P = spectral_decomposition_qubit(A)
-            p0 = self.ev(P[0], wires)  # probability of measuring a[0]
-            n0 = np.random.binomial(self.shots, p0)
-            ev = (n0*a[0] +(self.shots-n0)*a[1]) / self.shots
+            ev = np.mean(self.sample(observable, wires, par, self.shots))
 
         return ev
 
-    def var(self, expectation, wires, par):
-        # measurement/expectation value <psi|A|psi>
-        A = self._get_operator_matrix(expectation, par)
-        return self.ev(A@A, wires) - self.ev(A, wires)**2
+    def var(self, observable, wires, par):
+        if self.shots == 0:
+            # exact expectation value
+            A = self._get_operator_matrix(observable, par)
+            var = self.ev(A@A, wires) - self.ev(A, wires)**2
+        else:
+            # estimate the ev
+            var = np.var(self.sample(observable, wires, par, self.shots))
+
+        return var
+
+    def sample(self, observable, wires, par, n=None):
+        if n is None:
+            n = self.shots
+
+        if n == 0:
+            raise ValueError("Calling sample with n = 0 is not possible.")
+        if n < 0 or not isinstance(n, int):
+            raise ValueError("The number of samples must be a positive integer.")
+
+        A = self._get_operator_matrix(observable, par)
+        a, P = spectral_decomposition(A)
+
+        p = np.zeros(a.shape)
+        for idx, Pi in enumerate(P):
```

Comment:
worthwhile to use `map` here?
e.g.,
```
map_fn = lambda P: self.ev(P, wires)
p = list(map(map_fn, P))
```

Reply:
Nice solution. I didn't want to put too much effort into figuring out how to make numpy do it. I tried with `apply_along_axis` but that didn't work too well. 

But actually, as the matrix that is decomposed is represents an observable, we will only iterate over 2-4 values, so I don't think that using map is _necessary_ for performance reasons. In this case I would just keep the loop because it is very explicit